src/scenarios/shared/runner.py

```diff
@@ -9,38 +9,69 @@
 from collections import namedtuple
 from argparse import ArgumentParser
 from shared.startup import StartupWrapper
-from shared.util import publishedexe, extension
+from shared.util import publishedexe, extension, pythoncommand, iswin
 from shared import const
 from performance.logger import setup_loggers
 
-
-reqfields = ('exename',
-            )
-optfields = ('guiapp',
-             'startupmetric',
-             'appargs',
-             'environmentvariables',
-             'iterations',
-             'timeout',
-             'warmup',
-             'workingdir',
-             'iterationsetup',
-             'setupargs',
-             'iterationcleanup',
-             'cleanupargs',
-             'processwillexit',
-             'measurementdelay'
-             )
-
 # These are the kinds of scenarios we run. Default here indicates whether ALL
 # scenarios should try and run a given test type.
 testtypes = {const.STARTUP: False,
              const.SDK: False,
              const.CROSSGEN: False}
 
-TestTraits = namedtuple('TestTraits',
-                        reqfields  + tuple(testtypes.keys()) + optfields,
-                        defaults=tuple(testtypes.values()) + (None,) * len(optfields))
+
+class TestTraits:
+    traits = {}
+    # Scenario-independent traits that can be set in test.py,
+    # but can also be overriden in scenario-specific case
+    reqfields = ('exename',
+                 )
+    optfields = ('guiapp',
+                 'startupmetric',
+                 'appargs',
+                 'iterations',
+                 'timeout',
+                 'warmup',
+                 'workingdir',
+                 'iterationsetup',
+                 'setupargs',
+                 'iterationcleanup',
+                 'cleanupargs',
+                 'processwillexit',
+                 'measurementdelay',
+                 'environmentvariables'
+                 )
+
+    def __init__(self, **kwargs):
+        if 'exename' not in kwargs:
+            raise Exception("exename cannot be empty")
+        self.traits = dict.fromkeys(self.all_traits()) # initialize default traits
+        self.add_traits(**kwargs) # add initial traits
+
+    # add trait if not present or overwrite existing trait if overwrite=True
+    def add_trait(self, key: str, value: str, overwrite=True):
+        if not self.is_valid_trait(key):
+            raise Exception("%s is not a valid trait." % key)
+        if not self.get_trait(key) or overwrite:
+            self.traits[key] = value
+
+    # add multiple traits if not present or overwrite existing traits if overwrite=True
+    def add_traits(self, overwrite=True, **kwargs):
+        for keyword in kwargs:
+            self.add_trait(keyword, kwargs[keyword], overwrite=overwrite)
```

Comment:
`add_traits` feels weird because it reverses the overwrite param from the other one. I get that's necessary because of using `kwargs`. I wonder if we should just collapse these into a single method and just call `add_traits` always, even if there's only one trait? That would also be a nicer pattern when writing code, as you would unify on `trait=value` instead of `key=trait, value=value`.

Reply:
Totally reasonable!